### PR TITLE
feat: Set up eslint & prettier in es ds components and es ds docs

### PR DESCRIPTION
**Pull request by RANE1021**
### 🔗 Linked issue

- https://energysage.atlassian.net/browse/CED-1726

### ❓ Type of change

- [ ] 📖 Documentation (updates to the documentation or readme)
- [ ] 🐞 Bug fix (a non-breaking change that fixes an issue)
- [x] 👌 Enhancement (improving an existing functionality like performance)
- [ ] ✨ New feature (a non-breaking change that adds functionality)
- [ ] ⚠️ Breaking change (fix or feature that would cause existing functionality to change)

### 📚 Description

- Added individual linting for es docs and es components (this way the rules can deviate in the future)
- Fixed all major errors and some minor warnings ( see notes for extra comments)

### 🥼 Testing

- `npm run lint:js` in es-docs or es-components

#### 🧐 Feedback Requested / Focus Areas

- @tomleo What do you think we should do regarding the following? Change the overall linting rules? or just leave these as special cases we ignore?
  - /* eslint-disable no-unused-vars */
  - /* eslint-disable @typescript-eslint/no-unused-vars */
  - no-unused-vars
  - @typescript-eslint/no-unused-vars
  - /* eslint-disable vue/require-prop-types */ (in `es-collapse.vue` this line : `const model = defineModel();` )
  - /* eslint-disable no-use-before-define */ (`spacing.vue` example)
  - /* eslint-disable import/no-self-import */ (`spacing.vue` example)

### 📝 Checklist

- [ ] I have linked an issue or discussion.
- [ ] I have updated the documentation accordingly.
- [ ] I have documented testing approach

### CODE REVIEW

**Review by nathanielwarner — CHANGES_REQUESTED** [review]
I left some comments inline of things that should be addressed. Also, the lint command did not work for me out of the box, and I had to make a number of changes to get it working. I have these on this branch: https://github.com/EnergySage/es-ds/compare/CED-1726-Set-up-appropriate-eslint-prettier-in-es-ds-components-and-es-ds-docs...1726-more-linting-stuff

One notable thing on that branch is that eslint rules are updated to use the Vue 3 rules. This fixes a number of things like not recognizing the Nuxt autoimports.

Also noting that we haven't yet put in the rule for script -> template -> style ordering.

The branch I linked only gets us part of the way.. perhaps we should discuss approaches on this tomorrow?

File: es-ds-components/.eslintrc.json [hunk 1]
```diff
@@ -0,0 +1,133 @@
...
+        "parser": "@typescript-eslint/parser"
+    },
+    "settings": {
+        "import/resolver": {
+            "typescript": {
+                "alwaysTryTypes": true
+            },
+            "alias": {
+                "map": [
+                    ["@", "./lib"],
```
Comment (tomleo) [inline es-ds-components/.eslintrc.json:13]: This look like it was copied from https://github.com/EnergySage/es-storyblok-shared/blob/main/.eslintrc.json, I think the paths need to be tweaked for this repo

File: es-ds-components/.eslintrc.json [hunk 2]
```diff
@@ -0,0 +1,133 @@
...
+                    ["~", "."]
+                ]
+            }
+        }
+    },
+    "plugins": ["@typescript-eslint"],
```
Comment (tomleo) [inline es-ds-components/.eslintrc.json:19]: I wonder if something like `plugin:vue/vue3(-strongly)?-recommended` might be worth using, maybe:

File: es-ds-components/.eslintrc.json [hunk 3]
```diff
@@ -0,0 +1,133 @@
...
+    "overrides": [
+        {
+            "files": ["*.yaml", "*.yml"],
+            "parser": "yaml-eslint-parser",
+            "rules": {
+                "max-len": "off"
+            }
+        }
+    ],
+    "extends": [
```
Comment (tomleo) [inline es-ds-components/.eslintrc.json:32]: I wonder if we could start with something like:

```
  "extends": [
    "eslint:recommended",
    "plugin:@typescript-eslint/recommended",
    "plugin:vue/vue3-recommended",
    "prettier"
  ],
```

Per the docs https://eslint.vuejs.org/user-guide/#configuration-eslint-config-js

File: es-ds-components/.prettierrc.json [hunk 4]
```diff
@@ -0,0 +1,10 @@
+{
```
Comment (tomleo) [inline es-ds-components/.prettierrc.json:1]: 👍

**Review by tomleo — CHANGES_REQUESTED** [review]
Looks good, a few things inline 👍

**Comment by tomleo** [comment]
> What do you think we should do regarding the following? Change the overall linting rules? or just leave these as special cases we ignore?

For things like unused vars, it seems easy enough to remove the assignment. In general I'm thinking the less linting ignore comments the better, particularly in the design system (but I could be convinced otherwise).

**Comment by RANE1021** [comment]
> Looks good, a few things inline 👍

@tomleo take another look was in the middle of making and pushing changes as this came through

File: es-ds-components/.eslintrc.json (hunk 1 above)
Comment (RANE1021) [reply es-ds-components/.eslintrc.json]: @tomleo 
alias mapping causes this error in es docs, even after tweaked
`"alias": {
                "map": [
                    ["@", "../"]
                ]
            }`
removing it seems to get rid of that erorr.

also this error is the last remaining item
` error  Unable to resolve path to module '@energysage/es-ds-components/components/es-form-msg.vue?raw'  import/no-unresolved` not sure what this is.

File: es-ds-components/components/es-breadcrumbs.vue [hunk 5]
```diff
@@ -5,58 +5,71 @@
             menu: 'breadcrumb',
             menuitem: 'breadcrumb-item'
         }">
+        <!-- eslint-disable vue/no-unused-vars -->
+        <!-- eslint-disable vue/no-template-shadow -->
```
[the remark this answers is not in the document]
Comment (RANE1021) [reply es-ds-components/components/es-breadcrumbs.vue:34]: Needed to add globals nuxt vars to the eslint config

**Review by nathanielwarner — COMMENTED** [review]
@RANE1021 On the latest version of this PR branch, I'm not able to load things anymore and am getting this error:

Screenshot 2024-08-26 at 17 01 02

Is this something you're seeing too?

**Comment by RANE1021** [comment]
@nathanielwarner 
> Is this something you're seeing too?

 I just recently merged `esds-3.0-vue3-primevue` and I see the same error there. any ideas?

Screenshot 2024-08-27 at 9 32 12 AM

**Comment by nathanielwarner** [comment]
Hey @RANE1021! Could you resolve the merge conflicts and apply the linting to the latest changes? (It might be easiest to remove all the changes except the linting configs themselves and then re-lint)

File: es-ds-docs/layouts/default.vue
Comment (nathanielwarner) [inline es-ds-docs/layouts/default.vue:1]: This appears to be where the bad merge is

**Review by nathanielwarner — CHANGES_REQUESTED** [review]
It seems like the merge isn't correct, as the docs won't load and give this error: `ERROR  Internal server error: Single file component can contain only one <template> element`

**Review by nathanielwarner — APPROVED** [review]
All set! There's a followup issue to address which I'll make a ticket for- manually setting the Nuxt globals within the eslint config is not really workable, because there's many more of them, and there's also our own globals that get auto-imported from the composables folder, which also would need to be added to downstream repos.

**Pull request merged**